spec/mem/tree_borrows/memory.md

```diff
@@ -144,15 +144,14 @@ impl<T: Target> TreeBorrowsMemory<T> {
 
     /// Return the provenance of the pointer and offset of the pointer in the allocation.
     fn check_ptr(&self, ptr: Pointer<TreeBorrowsProvenance>, len: Size) -> Result<Option<(TreeBorrowsProvenance, Size)>> {
-        // For zero-sized accesses, there is nothing to check.
-        // (Provenance monotonicity says that if we allow zero-sized accesses
-        // for `None` provenance we have to allow it for all provenance.)
-        if len.is_zero() {
-            return ret(None);
-        }
         // We do not even have to check for null, since no allocation will ever contain that address.
         // Now try to access the allocation information.
         let Some((alloc_id, path)) = ptr.provenance else {
+            // We allow zero-sized accesses for `None` provenance.
+            // (Provenance monotonicity says that if we allow this,
+            // we have to allow it for all provenance.)
+            if len.is_zero() { return ret(None); }
```

Comment:
I understand now why you changed this, but it's not quite the right behavior. Probably retagging should not use check_ptr.

Please add two tests:
- it should be UB to violate the protector even with a zero-sized pointee, if it has provenance
- it should *not* be UB to have a zero-sized pointer that is outside the bounds of the allocation